File: backend/src/main/java/com/quantis/stock/service/ComptabiliteService.java
package com.quantis.stock.service;

import com.quantis.stock.dto.ClotureRequestDto;
import com.quantis.stock.dto.ClotureSyntheseDto;
import com.quantis.stock.dto.EcritureGrandLivreDto;
import com.quantis.stock.exception.ResourceNotFoundException;
import com.quantis.stock.model.*;
import com.quantis.stock.model.enums.StatutDocument;
import com.quantis.stock.model.enums.TypeCaisse;
import com.quantis.stock.model.enums.TypeDocument;
import com.quantis.stock.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.*;

/**
 * Service Comptabilité — Journal de Caisse, Grand Livre Général & Clôtures Périodiques.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class ComptabiliteService {

    private final MouvementCaisseRepository caisseRepository;
    private final UtilisateurRepository utilisateurRepository;
    private final SessionCaisseRepository sessionCaisseRepository;
    private final ClotureComptableRepository clotureRepository;
    private final DocumentRepository documentRepository;
    private final CommandeFournisseurRepository commandeFournisseurRepository;
    private final StockCourantRepository stockCourantRepository;
    private final ClientRepository clientRepository;
    private final FournisseurRepository fournisseurRepository;

    // =================== JOURNAL DE CAISSE ===================

    @Transactional
    public MouvementCaisse enregistrerMouvement(MouvementCaisse mouvement, String userEmail) {
        var utilisateur = utilisateurRepository.findByEmail(userEmail)
                .orElseThrow(() -> new ResourceNotFoundException("Utilisateur", "email", userEmail));
        mouvement.setUtilisateur(utilisateur);
        if (mouvement.getDateMouvement() == null) {
            mouvement.setDateMouvement(LocalDate.now());
        }

        // Lier automatiquement à la session de caisse active du caissier
        var sessionActive = sessionCaisseRepository.findByCaissierIdAndStatut(
                utilisateur.getId(), com.quantis.stock.model.enums.StatutSessionCaisse.OUVERTE);
        if (sessionActive.isPresent()) {
            var session = sessionActive.get();
            mouvement.setSessionCaisse(session);
            if (mouvement.getType() == TypeCaisse.ENTREE) {
                session.setTotalEntrees(session.getTotalEntrees().add(mouvement.getMontant()));
            } else if (mouvement.getType() == TypeCaisse.SORTIE) {
                session.setTotalSorties(session.getTotalSorties().add(mouvement.getMontant()));
            }
            session.recalculerSoldeTheorique();
            sessionCaisseRepository.save(session);
        }

        mouvement = caisseRepository.save(mouvement);
        log.info("Mouvement caisse {}: {} FCFA — {}", mouvement.getType(), mouvement.getMontant(), mouvement.getLibelle());
        return mouvement;
    }

    @Transactional(readOnly = true)
    public Page<MouvementCaisse> getJournal(LocalDate debut, LocalDate fin, Pageable pageable) {
        return caisseRepository.findByDateMouvementBetweenOrderByDateMouvementDesc(debut, fin, pageable);
    }

    // =================== GRAND LIVRE GÉNÉRAL ===================

    @Transactional(readOnly = true)
    public List<EcritureGrandLivreDto> getGrandLivre(LocalDate debut, LocalDate fin) {
        List<EcritureGrandLivreDto> ecritures = new ArrayList<>();

        // 1. Factures de Vente Validées (Ventes de marchandises / Produits d'exploitation)
        List<Document> factures = documentRepository.findByDateDocumentBetweenAndStatut(
                debut, fin, StatutDocument.VALIDE);
        for (Document f : factures) {
            if (f.getType() == TypeDocument.FACTURE || f.getType() == TypeDocument.BON_LIVRAISON) {
                String clientNom = f.getClient() != null ? f.getClient().getNom() : "Client Comptoir";
                ecritures.add(EcritureGrandLivreDto.builder()
                        .date(f.getDateDocument() != null ? f.getDateDocument() : LocalDate.now())
                        .typeFlux("VENTE")
                        .referencePiece(f.getNumero())
                        .tiersOuCategorie(clientNom)
                        .libelle("Vente facturée " + f.getNumero() + " (" + clientNom + ")")
                        .debit(f.getTotalTtc()) // Débit client / Chiffre d'affaires
                        .credit(BigDecimal.ZERO)
                        .build());
            }
        }

        // 2. Commandes d'Achat Fournisseurs (Charges / Achats de marchandises)
        List<CommandeFournisseur> achats = commandeFournisseurRepository.findByDateCommandeBetween(debut, fin);
        for (CommandeFournisseur a : achats) {
            if (a.getStatut() != com.quantis.stock.model.enums.StatutCommande.ANNULEE
                    && a.getStatut() != com.quantis.stock.model.enums.StatutCommande.BROUILLON) {
                String fourNom = a.getFournisseur() != null ? a.getFournisseur().getNom() : "Fournisseur";
                ecritures.add(EcritureGrandLivreDto.builder()
                        .date(a.getDateCommande() != null ? a.getDateCommande() : LocalDate.now())
                        .typeFlux("ACHAT")
                        .referencePiece(a.getNumero())
                        .tiersOuCategorie(fourNom)
                        .libelle("Achat stock " + a.getNumero() + " (" + fourNom + ")")
                        .debit(BigDecimal.ZERO)
                        .credit(a.getTotalHt() != null ? a.getTotalHt() : BigDecimal.ZERO) // Crédit fournisseur
                        .build());
            }
        }

        // 3. Mouvements de Caisse (Encaissements & Décaissements effectifs)
        List<MouvementCaisse> mouvements = caisseRepository.findByDateMouvementBetweenOrderByDateMouvementAsc(debut, fin);
        for (MouvementCaisse m : mouvements) {
            boolean isEntree = m.getType() == TypeCaisse.ENTREE;
            ecritures.add(EcritureGrandLivreDto.builder()
                    .date(m.getDateMouvement())
                    .typeFlux(isEntree ? "ENCAISSEMENT_CAISSE" : "DEPENSE_CAISSE")
                    .referencePiece(m.getReference() != null && !m.getReference().isBlank() ? m.getReference() : "MC-" + m.getId())
                    .tiersOuCategorie(m.getCategorie() != null ? m.getCategorie() : (isEntree ? "Recettes" : "Frais Généraux"))
                    .libelle(m.getLibelle())
                    .debit(isEntree ? m.getMontant() : BigDecimal.ZERO)
                    .credit(isEntree ? BigDecimal.ZERO : m.getMontant())
                    .build());
        }

        // Tri chronologique ascendant
        ecritures.sort(Comparator.comparing(EcritureGrandLivreDto::getDate));

        // Calcul du solde progressif
        BigDecimal soldeCumule = BigDecimal.ZERO;
        for (EcritureGrandLivreDto e : ecritures) {
            soldeCumule = soldeCumule.add(e.getDebit()).subtract(e.getCredit());
            e.setSoldeProgressif(soldeCumule);
        }

        return ecritures;
    }

    // =================== CLÔTURES PÉRIODIQUES ===================

    @Transactional(readOnly = true)
    public ClotureSyntheseDto simulerCloture(LocalDate debut, LocalDate fin, String periode) {
        // 1. Chiffre d'affaires & TVA
        List<Document> factures = documentRepository.findByDateDocumentBetweenAndStatut(
                debut, fin, StatutDocument.VALIDE);
        BigDecimal caTtc = BigDecimal.ZERO;
        BigDecimal caHt = BigDecimal.ZERO;
        BigDecimal tvaCollectee = BigDecimal.ZERO;
        int nbVentes = 0;

        for (Document f : factures) {
            if (f.getType() == TypeDocument.FACTURE || f.getType() == TypeDocument.BON_LIVRAISON) {
                caTtc = caTtc.add(f.getTotalTtc() != null ? f.getTotalTtc() : BigDecimal.ZERO);
                caHt = caHt.add(f.getTotalHt() != null ? f.getTotalHt() : BigDecimal.ZERO);
                tvaCollectee = tvaCollectee.add(f.getTotalTva() != null ? f.getTotalTva() : BigDecimal.ZERO);
                nbVentes++;
            }
        }

        // 2. Achats fournisseurs
        List<CommandeFournisseur> achats = commandeFournisseurRepository.findByDateCommandeBetween(debut, fin);
        BigDecimal achatsHt = BigDecimal.ZERO;
        int nbAchats = 0;

        for (CommandeFournisseur a : achats) {
            if (a.getStatut() != com.quantis.stock.model.enums.StatutCommande.ANNULEE) {
                achatsHt = achatsHt.add(a.getTotalHt() != null ? a.getTotalHt() : BigDecimal.ZERO);
                nbAchats++;
            }
        }

        // 3. Mouvements de Caisse
        BigDecimal totalEntrees = caisseRepository.sumByTypeAndPeriode(TypeCaisse.ENTREE, debut, fin);
        BigDecimal totalSorties = caisseRepository.sumByTypeAndPeriode(TypeCaisse.SORTIE, debut, fin);
        if (totalEntrees == null) totalEntrees = BigDecimal.ZERO;
        if (totalSorties == null) totalSorties = BigDecimal.ZERO;
        BigDecimal soldeCaisse = totalEntrees.subtract(totalSorties);

        // 4. Marge brute estimée
        BigDecimal margeBrute = caHt.subtract(achatsHt);

        // 5. Valorisation du Stock final
        List<StockCourant> stocks = stockCourantRepository.findAll();
        BigDecimal valeurStock = BigDecimal.ZERO;
        for (StockCourant sc : stocks) {
            if (sc.getQuantite() != null && sc.getQuantite().compareTo(BigDecimal.ZERO) > 0) {
                BigDecimal prix = sc.getProduit().getPrixAchat() != null
                        ? sc.getProduit().getPrixAchat()
                        : (sc.getProduit().getPrixVente() != null ? sc.getProduit().getPrixVente() : BigDecimal.ZERO);
                valeurStock = valeurStock.add(sc.getQuantite().multiply(prix));
            }
        }

        // 6. Créances & Dettes
        BigDecimal creances = clientRepository.sumSoldeCredit();
        if (creances == null) creances = BigDecimal.ZERO;
        BigDecimal dettes = fournisseurRepository.sumSoldeDette();
        if (dettes == null) dettes = BigDecimal.ZERO;

        boolean dejaCloturee = clotureRepository.existsByPeriode(periode);

        return ClotureSyntheseDto.builder()
                .periode(periode)
                .libelle("Bilan Périodique — " + periode)
                .dateDebut(debut)
                .dateFin(fin)
                .chiffreAffairesTtc(caTtc)
                .chiffreAffairesHt(caHt)
                .totalTvaCollectee(tvaCollectee)
                .totalAchatsHt(achatsHt)
                .totalDepensesCaisse(totalSorties)
                .totalEntreesCaisse(totalEntrees)
                .margeBruteEstimee(margeBrute)
                .valeurStockFinPeriode(valeurStock)
                .totalCreancesClients(creances)
                .totalDettesFournisseurs(dettes)
                .soldeCaisseFinal(soldeCaisse)
                .nombreVentes(nbVentes)
                .nombreAchats(nbAchats)
                .dejaCloturee(dejaCloturee)
                .build();
    }

    @Transactional
    public ClotureComptable validerCloture(ClotureRequestDto req, String userEmail) {
        var utilisateur = utilisateurRepository.findByEmail(userEmail)
                .orElseThrow(() -> new ResourceNotFoundException("Utilisateur", "email", userEmail));

        // Vérifier si déjà clôturé
        clotureRepository.findByPeriode(req.getPeriode()).ifPresent(c -> {
            throw new IllegalArgumentException("La période « " + req.getPeriode() + " » a déjà fait l'objet d'une clôture verrouillée.");
        });

        ClotureSyntheseDto synthese = simulerCloture(req.getDateDebut(), req.getDateFin(), req.getPeriode());

        ClotureComptable cloture = ClotureComptable.builder()
                .periode(req.getPeriode())
                .libelle(req.getLibelle() != null && !req.getLibelle().isBlank() ? req.getLibelle() : "Arrêté Mensuel " + req.getPeriode())
                .dateDebut(req.getDateDebut())
                .dateFin(req.getDateFin())
                .dateCloture(Instant.now())
                .cloturePar(utilisateur)
                .chiffreAffairesTtc(synthese.getChiffreAffairesTtc())
                .chiffreAffairesHt(synthese.getChiffreAffairesHt())
                .totalTvaCollectee(synthese.getTotalTvaCollectee())
                .totalAchatsHt(synthese.getTotalAchatsHt())
                .totalDepensesCaisse(synthese.getTotalDepensesCaisse())
                .totalEntreesCaisse(synthese.getTotalEntreesCaisse())
                .margeBruteEstimee(synthese.getMargeBruteEstimee())
                .valeurStockFinPeriode(synthese.getValeurStockFinPeriode())
                .totalCreancesClients(synthese.getTotalCreancesClients())
                .totalDettesFournisseurs(synthese.getTotalDettesFournisseurs())
                .soldeCaisseFinal(synthese.getSoldeCaisseFinal())
                .statut("VERROUILLE")
                .notes(req.getNotes())
                .build();

        cloture = clotureRepository.save(cloture);
        log.info("Clôture comptable validée et verrouillée pour la période {}: CA TTC = {} FCFA par {}",
                cloture.getPeriode(), cloture.getChiffreAffairesTtc(), userEmail);

        return cloture;
    }

    @Transactional(readOnly = true)
    public List<ClotureComptable> getHistoriqueClotures() {
        return clotureRepository.findAllByOrderByDateClotureDesc();
    }

    // =================== RAPPORTS SIMPLES ===================

    @Transactional(readOnly = true)
    public Map<String, Object> getRapportPeriode(LocalDate debut, LocalDate fin) {
        BigDecimal totalEntrees = caisseRepository.sumByTypeAndPeriode(TypeCaisse.ENTREE, debut, fin);
        BigDecimal totalSorties = caisseRepository.sumByTypeAndPeriode(TypeCaisse.SORTIE, debut, fin);
        if (totalEntrees == null) totalEntrees = BigDecimal.ZERO;
        if (totalSorties == null) totalSorties = BigDecimal.ZERO;
        BigDecimal solde = totalEntrees.subtract(totalSorties);

        return Map.of(
                "debut", debut.toString(),
                "fin", fin.toString(),
                "totalEntrees", totalEntrees,
                "totalSorties", totalSorties,
                "solde", solde
        );
    }
}

File: frontend/lib/features/comptabilite/data/comptabilite_service.dart
import 'package:dio/dio.dart';
import '../../../core/network/api_client.dart';
import '../../../core/utils/file_download_helper.dart';

class EcritureGrandLivre {
  final String date;
  final String typeFlux;
  final String referencePiece;
  final String tiersOuCategorie;
  final String libelle;
  final double debit;
  final double credit;
  final double soldeProgressif;

  EcritureGrandLivre({
    required this.date,
    required this.typeFlux,
    required this.referencePiece,
    required this.tiersOuCategorie,
    required this.libelle,
    required this.debit,
    required this.credit,
    required this.soldeProgressif,
  });

  factory EcritureGrandLivre.fromJson(Map<String, dynamic> json) {
    return EcritureGrandLivre(
      date: json['date']?.toString() ?? '',
      typeFlux: json['typeFlux']?.toString() ?? '',
      referencePiece: json['referencePiece']?.toString() ?? '',
      tiersOuCategorie: json['tiersOuCategorie']?.toString() ?? '',
      libelle: json['libelle']?.toString() ?? '',
      debit: (json['debit'] as num? ?? 0).toDouble(),
      credit: (json['credit'] as num? ?? 0).toDouble(),
      soldeProgressif: (json['soldeProgressif'] as num? ?? 0).toDouble(),
    );
  }
}

class ClotureSynthese {
  final String periode;
  final String libelle;
  final String dateDebut;
  final String dateFin;
  final double chiffreAffairesTtc;
  final double chiffreAffairesHt;
  final double totalTvaCollectee;
  final double totalAchatsHt;
  final double totalDepensesCaisse;
  final double totalEntreesCaisse;
  final double margeBruteEstimee;
  final double valeurStockFinPeriode;
  final double totalCreancesClients;
  final double totalDettesFournisseurs;
  final double soldeCaisseFinal;
  final int nombreVentes;
  final int nombreAchats;
  final bool dejaCloturee;

  ClotureSynthese({
    required this.periode,
    required this.libelle,
    required this.dateDebut,
    required this.dateFin,
    required this.chiffreAffairesTtc,
    required this.chiffreAffairesHt,
    required this.totalTvaCollectee,
    required this.totalAchatsHt,
    required this.totalDepensesCaisse,
    required this.totalEntreesCaisse,
    required this.margeBruteEstimee,
    required this.valeurStockFinPeriode,
    required this.totalCreancesClients,
    required this.totalDettesFournisseurs,
    required this.soldeCaisseFinal,
    required this.nombreVentes,
    required this.nombreAchats,
    required this.dejaCloturee,
  });

  factory ClotureSynthese.fromJson(Map<String, dynamic> json) {
    return ClotureSynthese(
      periode: json['periode']?.toString() ?? '',
      libelle: json['libelle']?.toString() ?? '',
      dateDebut: json['dateDebut']?.toString() ?? '',
      dateFin: json['dateFin']?.toString() ?? '',
      chiffreAffairesTtc: (json['chiffreAffairesTtc'] as num? ?? 0).toDouble(),
      chiffreAffairesHt: (json['chiffreAffairesHt'] as num? ?? 0).toDouble(),
      totalTvaCollectee: (json['totalTvaCollectee'] as num? ?? 0).toDouble(),
      totalAchatsHt: (json['totalAchatsHt'] as num? ?? 0).toDouble(),
      totalDepensesCaisse: (json['totalDepensesCaisse'] as num? ?? 0).toDouble(),
      totalEntreesCaisse: (json['totalEntreesCaisse'] as num? ?? 0).toDouble(),
      margeBruteEstimee: (json['margeBruteEstimee'] as num? ?? 0).toDouble(),
      valeurStockFinPeriode: (json['valeurStockFinPeriode'] as num? ?? 0).toDouble(),
      totalCreancesClients: (json['totalCreancesClients'] as num? ?? 0).toDouble(),
      totalDettesFournisseurs: (json['totalDettesFournisseurs'] as num? ?? 0).toDouble(),
      soldeCaisseFinal: (json['soldeCaisseFinal'] as num? ?? 0).toDouble(),
      nombreVentes: json['nombreVentes'] as int? ?? 0,
      nombreAchats: json['nombreAchats'] as int? ?? 0,
      dejaCloturee: json['dejaCloturee'] as bool? ?? false,
    );
  }
}

class ClotureComptableModel {
  final int id;
  final String periode;
  final String libelle;
  final String dateDebut;
  final String dateFin;
  final String dateCloture;
  final String clotureParNom;
  final double chiffreAffairesTtc;
  final double margeBruteEstimee;
  final double soldeCaisseFinal;
  final double valeurStockFinPeriode;
  final String statut;
  final String? notes;

  ClotureComptableModel({
    required this.id,
    required this.periode,
    required this.libelle,
    required this.dateDebut,
    required this.dateFin,
    required this.dateCloture,
    required this.clotureParNom,
    required this.chiffreAffairesTtc,
    required this.margeBruteEstimee,
    required this.soldeCaisseFinal,
    required this.valeurStockFinPeriode,
    required this.statut,
    this.notes,
  });

  factory ClotureComptableModel.fromJson(Map<String, dynamic> json) {
    return ClotureComptableModel(
      id: json['id'] as int? ?? 0,
      periode: json['periode']?.toString() ?? '',
      libelle: json['libelle']?.toString() ?? '',
      dateDebut: json['dateDebut']?.toString() ?? '',
      dateFin: json['dateFin']?.toString() ?? '',
      dateCloture: json['dateCloture']?.toString() ?? '',
      clotureParNom: json['cloturePar']?['nomComplet']?.toString() ?? 'Administrateur',
      chiffreAffairesTtc: (json['chiffreAffairesTtc'] as num? ?? 0).toDouble(),
      margeBruteEstimee: (json['margeBruteEstimee'] as num? ?? 0).toDouble(),
      soldeCaisseFinal: (json['soldeCaisseFinal'] as num? ?? 0).toDouble(),
      valeurStockFinPeriode: (json['valeurStockFinPeriode'] as num? ?? 0).toDouble(),
      statut: json['statut']?.toString() ?? 'VERROUILLE',
      notes: json['notes']?.toString(),
    );
  }
}

class ComptabiliteService {
  final Dio _dio = ApiClient.instance;

  Future<List<EcritureGrandLivre>> getGrandLivre({
    required DateTime debut,
    required DateTime fin,
  }) async {
    final res = await _dio.get(
      '/accounting/grand-livre',
      queryParameters: {
        'debut': debut.toIso8601String().substring(0, 10),
        'fin': fin.toIso8601String().substring(0, 10),
      },
    );
    final list = res.data['data'] as List? ?? [];
    return list.map((e) => EcritureGrandLivre.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<ClotureSynthese> simulerCloture({
    required DateTime debut,
    required DateTime fin,
    required String periode,
  }) async {
    final res = await _dio.get(
      '/accounting/clotures/simuler',
      queryParameters: {
        'debut': debut.toIso8601String().substring(0, 10),
        'fin': fin.toIso8601String().substring(0, 10),
        'periode': periode,
      },
    );
    return ClotureSynthese.fromJson(res.data['data'] as Map<String, dynamic>);
  }

  Future<ClotureComptableModel> validerCloture({
    required String periode,
    String? libelle,
    required DateTime debut,
    required DateTime fin,
    String? notes,
  }) async {
    final res = await _dio.post(
      '/accounting/clotures',
      data: {
        'periode': periode,
        'libelle': libelle ?? 'Arrêté Mensuel $periode',
        'dateDebut': debut.toIso8601String().substring(0, 10),
        'dateFin': fin.toIso8601String().substring(0, 10),
        'notes': notes ?? '',
      },
    );
    return ClotureComptableModel.fromJson(res.data['data'] as Map<String, dynamic>);
  }

  Future<List<ClotureComptableModel>> getHistoriqueClotures() async {
    final res = await _dio.get('/accounting/clotures');
    final list = res.data['data'] as List? ?? [];
    return list.map((e) => ClotureComptableModel.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<void> exportGrandLivreCsv({
    required DateTime debut,
    required DateTime fin,
  }) async {
    final debutStr = debut.toIso8601String().substring(0, 10);
    final finStr = fin.toIso8601String().substring(0, 10);
    final res = await _dio.get(
      '/accounting/grand-livre/export',
      queryParameters: {'debut': debutStr, 'fin': finStr},
      options: Options(responseType: ResponseType.bytes),
    );
    FileDownloadHelper.download(
      res.data as List<int>,
      'grand_livre_${debutStr}_$finStr.csv',
      mimeType: 'text/csv;charset=utf-8',
    );
  }

  Future<Map<String, dynamic>> seedDemoData() async {
    final res = await _dio.post('/accounting/seed-demo');
    return res.data['data'] as Map<String, dynamic>? ?? {};
  }
}
